src/afrolabs/components/kafka/checkpoint_storage/stores.clj

```diff
@@ -0,0 +1,684 @@
+  (let [bs (byte-array *chunk-size-bytes*)]
+    (loop [remaining-bytes *chunk-size-bytes*]
+      (let [read-this-time (.read input-stream
+                                  bs
+                                  (- *chunk-size-bytes* remaining-bytes)
+                                  remaining-bytes)
+            loop-step-result
+            (cond
+              ;; the stream is done, call back with what we have
+              (= -1 read-this-time)
+              (let [in-this-chunk (- *chunk-size-bytes* remaining-bytes)]
+                (when (pos? in-this-chunk)
+                    (callback-with-chunk bs in-this-chunk))
+                  nil ;; we are done uploading chunks
+                  )
+
+              ;; we read as much as we had space, we have a full chunk
+              (= remaining-bytes read-this-time)
+              (do (callback-with-chunk bs *chunk-size-bytes*)
+                  *chunk-size-bytes*)
+
+              ;; we read less than remaining-bytes, read some more, the chunk has more space left
+              (< read-this-time remaining-bytes)
+              (- remaining-bytes read-this-time))]
+
+        (when loop-step-result
+          (recur loop-step-result))))
+    (callback-with-chunk nil -1)))
+
+(comment
+
+  (binding [*chunk-size-bytes* 2]
+    (let [bs  (byte-array (range 25))
+          result-a (atom [])]
+      (with-open [bis (java.io.ByteArrayInputStream. bs)]
+        (input-stream->chunks bis
+                              #(when %1 ;; on the last call, byte-array is `nil` and number-of-bytes is `-1`
+                                 (swap! result-a
+                                        into
+                                        ;; important that a copy is made of the incoming byte-array contents
+                                        (take %2 %1)))))
+      @result-a))
+  ;; [0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 21 22 23 24]
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+  (require '[afrolabs.components.aws.sso :as -aws-sso-profile-provider])
+
+  (def s3-client (aws/client {:api :s3
+                              :region "af-south-1"
+                              :credentials-provider (-aws-sso-profile-provider/provider (or (System/getenv "AWS_PROFILE")
+                                                                                            (System/getProperty "aws.profile")
+                                                                                            "default"))}))
+
+  (aws/ops s3-client)
+  (aws/invoke s3-client
+              {:op :ListObjects
+               :request {:Bucket "ktable-checkpoint-store20250401120818030200000001"}})
+
+
+  )
+
+;;;;;;;;;;;;;;;;;;;;
+
+(defn- input-stream->multi-part-s3-upload
+  "Accepts a (java.io.) InputStream, an s3-client and an s3 object coordinate/destination.
+
+  Breaks up the InputStream into chunks of size `*chunk-size-bytes*`
+  and uploads each chunk using S3 Multipart upload.
```

Comment:
Make sure to add a lifecycle policy to the bucket to delete incomplete multipart uploads https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpu-abort-incomplete-mpu-lifecycle-config.html